llap-server/src/java/org/apache/hadoop/hive/llap/cache/ClockCachePolicy.java

```diff
@@ -0,0 +1,264 @@
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Clock eviction policy. Uses a simple circular list to keep a ring of current used buffers.
+ * New entries are added to tail of the clock hand AKA (clockHand.prev)
+ * Eviction start at the current clock hand following the next pointer.
+ *
+ */
+public class ClockCachePolicy implements LowLevelCachePolicy {
+
+  private static final int DEFAULT_MAX_CIRCLES = 5;
+
+  /**
+   * Lock to protect the state of the policy, used as mutex when modifying the circular linked list.
+   */
+  private final Lock listLock = new ReentrantLock();
+
+  /**
+   * The clock hand shared between threads thus made volatile, to ensure state when read outside of lock.
+   */
+  @GuardedBy("listLock")
+  private volatile LlapCacheableBuffer clockHand;
+
+  private EvictionListener evictionListener;
+  /**
+   * Max number of clock rotation before giving up on clock operation like eviction.
+   */
+  private final int maxCircles;
+
+  public ClockCachePolicy() {
+    maxCircles = DEFAULT_MAX_CIRCLES;
+  }
+
+  public ClockCachePolicy(int maxCircles) {
+    Preconditions.checkState(maxCircles > 0, "Maximum number of clock rotation must be positive and got " + maxCircles);
+    this.maxCircles = maxCircles;
+  }
+
+  /**
+   * Signals to the policy the addition of a new entry to the cache. An entry come with a priority that can be used as
+   * a hint to replacement policy.
+   *
+   * @param buffer   buffer to be cached
+   * @param priority the priority of cached element
+   */
+  @Override public void cache(LlapCacheableBuffer buffer, LowLevelCache.Priority priority) {
+    listLock.lock();
+    try {
+      clockHand = appendToCircularList(clockHand, buffer);
+    } finally {
+      listLock.unlock();
+    }
+  }
+
+  /**
+   * Appends new entry to the tail of circular list.
+   *
+   * @param head   circular list head.
+   * @param buffer new entry to be added.
+   * @return the ring head.
+   */
+  private static LlapCacheableBuffer appendToCircularList(LlapCacheableBuffer head, LlapCacheableBuffer buffer) {
+    if (head == null) {
+      return linkToItSelf(buffer);
+    }
+    buffer.next = head;
+    buffer.prev = head.prev;
+    head.prev.next = buffer;
+    head.prev = buffer;
+    return head;
+  }
+
+  /**
+   * Links the entry to it self to form a ring.
+   *
+   * @param buffer input
+   * @return buffer
+   */
+  private static LlapCacheableBuffer linkToItSelf(LlapCacheableBuffer buffer) {
+    buffer.prev = buffer;
+    buffer.next = buffer;
+    return buffer;
+  }
+
+  @Override public void notifyLock(LlapCacheableBuffer buffer) {
+    buffer.setClockBit();
+  }
+
+  /**
+   * Notifies the policy that a buffer is unlocked after been used. This notification signals to the policy that an
+   * access to this page occurred.
+   *
+   * @param buffer buffer that just got unlocked after a read.
+   */
+  @Override public void notifyUnlock(LlapCacheableBuffer buffer) {
```

Comment:
The bit set is lock free so i do not think will change much TBH. to avoid changing lot of the tests setup i will keep it as is and change it in the future if we see a strong case for it. Thanks